### PR TITLE
applications: nrf_desktop: Sending keyboard event via USB

### PR DESCRIPTION
**Pull request by MarekPieta**
Change adds sending hid keyboard events to USB subscribers.

### CODE REVIEW

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 1]
```diff
@@ -23,28 +24,24 @@
...
 
-	if (err || !atomic_get(&usb_ready)) {
-		return BT_GATT_ITER_CONTINUE;
-	}
+struct keyboard_event_item {
+	sys_snode_t node;
+	struct hid_keyboard_event *evt;
+};
 
+sys_slist_t keyboard_event_list;
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_forward.c]: static

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 2]
```diff
@@ -23,28 +24,24 @@
...
-{
-	if (!data) {
-		return BT_GATT_ITER_STOP;
-	}
+static struct hid_mouse_event *next_mouse_event;
 
-	if (err || !atomic_get(&usb_ready)) {
-		return BT_GATT_ITER_CONTINUE;
-	}
+struct keyboard_event_item {
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_forward.c]: define types before declaring variables

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 3]
```diff
@@ -23,28 +24,24 @@
 #include <logging/log.h>
 LOG_MODULE_REGISTER(MODULE, CONFIG_DESKTOP_BLE_SCANNING_LOG_LEVEL);
 
-
-static struct bt_gatt_hids_c hidc;
+static struct bt_gatt_hids_c hidc_arr[CONFIG_BT_MAX_CONN];
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_forward.c]: simply `hidc` or `hidc_instance`

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 4]
```diff
@@ -86,10 +83,65 @@ static u8_t hidc_read(struct bt_gatt_hids_c *hids_c,
...
+
+static void process_keyboard_report(const u8_t *data)
+{
+	struct hid_keyboard_event *event = new_hid_keyboard_event();
+
+	event->subscriber = usb_id;
+	event->modifier_bm = data[0];
+	memcpy(event->keys, &data[2], ARRAY_SIZE(event->keys));
+
+	u32_t key = irq_lock();
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_forward.c]: please convert these irq_lock calls to spinlock

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 5]
```diff
@@ -86,10 +83,65 @@ static u8_t hidc_read(struct bt_gatt_hids_c *hids_c,
...
+
+	if (!usb_busy) {
+		EVENT_SUBMIT(event);
+		usb_busy = true;
+	} else {
+		struct keyboard_event_item *evt_item = k_malloc(sizeof(*evt_item));
+
+		if (!evt_item) {
+			LOG_PANIC();
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_forward.c]: k_panic maybe?

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 6]
```diff
@@ -86,10 +83,65 @@ static u8_t hidc_read(struct bt_gatt_hids_c *hids_c,
...
+	event->modifier_bm = data[0];
+	memcpy(event->keys, &data[2], ARRAY_SIZE(event->keys));
+
+	u32_t key = irq_lock();
+
+	if (!usb_busy) {
+		EVENT_SUBMIT(event);
+		usb_busy = true;
+	} else {
+		struct keyboard_event_item *evt_item = k_malloc(sizeof(*evt_item));
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_forward.c]: don't do malloc under a lock

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 7]
```diff
@@ -173,9 +236,9 @@ void hidc_write_cb(struct bt_gatt_hids_c *hidc,
 	}
 }
 
-u8_t hidc_read_cb(struct bt_gatt_hids_c *hidc,
-		  struct bt_gatt_hids_c_rep_info *rep,
-		  u8_t err, const u8_t *data)
+u8_t hidc_read_cfg(struct bt_gatt_hids_c *hidc,
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_forward.c]: static

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 8]
```diff
@@ -217,9 +280,18 @@ static bool event_handler(const struct event_header *eh)
...
-			next_event = NULL;
+		sys_snode_t *kbd_event_node = sys_slist_get(&keyboard_event_list);
+
+		if (kbd_event_node) {
+			struct keyboard_event_item *next_item =
+				CONTAINER_OF(kbd_event_node,
+					     struct keyboard_event_item, node);
+
+			EVENT_SUBMIT(next_item->evt);
+			k_free(next_item);
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_forward.c]: release memory after lock is released

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 1 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c]: done

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 2 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c]: done

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 3 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c]: renamed to `hidc`

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 7 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c]: done

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 5 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c]: done

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 4 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c]: done

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 6 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c]: ok

File: applications/nrf_desktop/src/modules/hid_forward.c (hunk 8 above)
Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c]: done

File: applications/nrf_desktop/src/modules/hid_forward.c [hunk 9]
```diff
@@ -216,14 +279,24 @@ u8_t hidc_read_cb(struct bt_gatt_hids_c *hidc,
 static bool event_handler(const struct event_header *eh)
 {
 	if (is_hid_report_sent_event(eh)) {
-		u32_t key = irq_lock();
-		if (next_event) {
-			EVENT_SUBMIT(next_event);
-			next_event = NULL;
+		k_spinlock_key_t key = k_spin_lock(&spinlock);
+		sys_snode_t *kbd_event_node = sys_slist_get(&keyboard_event_list);
+		struct keyboard_event_item *next_item = NULL;
```
Comment (pdunaj) [inline applications/nrf_desktop/src/modules/hid_forward.c]: move `struct keyboard_event_item *next_item = NULL;` before the lock is taken to indicate that it is used outside the lock

Comment (MarekPieta) [reply applications/nrf_desktop/src/modules/hid_forward.c]: done

**Pull request merged**